File: board.py
SIZE = 9
KOMI = 6.5

EMPTY = 0
BLACK = 1
WHITE = 2

TONARI = None


def _load_tonari():
    global TONARI
    TONARI = [None] * (SIZE * SIZE)
    for i in xrange(SIZE):
        for j in xrange(SIZE):
            tonari = []
            if i != 0:
                tonari.append((i - 1) * SIZE + j)
            if j != 0:
                tonari.append(i * SIZE + (j - 1))
            if i != SIZE - 1:
                tonari.append((i + 1) * SIZE + j)
            if i != SIZE - 1:
                tonari.append(i * SIZE + (j + 1))
            TONARI[i * SIZE + j] = tonari
_load_tonari()


class Game(object):
    def __init__(self):
        self.board = [EMPTY] * (SIZE * SIZE)
        self.captures = [0, 0]

    def has_lib(self, loc):
        init_color = self.board[loc]
        if init_color == EMPTY:
            return True

        frontier = []
        inspected = set([loc])

        while True:
            for tonari in TONARI[loc]:
                color = self.board[tonari]
                if color == 0:
                    return True
                elif color == init_color:
                    if tonari not in inspected:
                        frontier.append(tonari)
            if not frontier:
                return False
            loc = frontier.pop()
            inspected.add(loc)

    def kill_group(self, loc):
        init_color = self.board[loc]
        if init_color == EMPTY:
            return 0

        frontier = []
        inspected = set([loc])
        killed = 0

        while True:
            self.board[loc] = EMPTY
            killed += 1
            for tonari in TONARI[loc]:
                color = self.board[tonari]
                if color == init_color:
                    if tonari not in inspected:
                        frontier.append(tonari)
            if not frontier:
                return killed
            loc = frontier.pop()
            inspected.add(loc)

    def play(self, pos, color):
        x, y = pos
        loc = x * SIZE + y
        if self.board[loc] != EMPTY:
            return False

        self.board[loc] = color

        killing = []
        for tonari in TONARI[loc]:
            if color == 3 - self.board[tonari] and not self.has_lib(tonari):
                killing.append(tonari)

        if not any(killing) and not self.has_lib(loc):
            self.board[loc] = EMPTY
            return False

        for tonari in killing:
            self.captures[color - 1] += self.kill_group(tonari)

        return True


    def __str__(self):
        s = ''
        for i in xrange(SIZE):
            for j in xrange(SIZE):
                color = self.board[i * SIZE + j]
                if color == 0:
                    if i in (2, 6) and j in (2, 6):
                        s += '+'
                    else:
                        s += '.'
                elif color == 1:
                    s += 'X'
                elif color == 2:
                    s += 'O'
                s += ' '
            s += '\n'
        s += "X:%d, O:%d+%s" % (self.captures[0], self.captures[1], KOMI)
        return s

def test():
    b = Game()
    b.play((0, 0), BLACK);
    b.play((0, 2), BLACK);

    b.play((1, 0), WHITE);
    b.play((1, 2), WHITE);
    b.play((0, 3), WHITE);
    b.play((0, 1), WHITE);

    b.play((1, 1), BLACK);
    b.play((2, 0), BLACK);
    b.play((0, 0), BLACK);
    b.play((1, 0), BLACK);

    b.play((2, 1), WHITE);
    b.play((3, 0), WHITE);

    #print b

if __name__ == '__main__':
    import time
    a = time.time()
    for i in xrange(1000000):
        test()
    b = time.time()
    print "TIME:", b - a

File: board.c
#include "board.h"

Pos TONARI[SIZE * SIZE][5];
Pos DIAGONALS[SIZE * SIZE][5];
Arena * BOARD_ARENA;

int board_initialize() {
    int i, j;
    Pos * tonari;
    LOOP(i, SIZE) {
        LOOP(j, SIZE) {
            tonari = TONARI[to_pos(i, j)];
            if(i != 0) *(tonari++) = to_pos(i - 1, j);
            if(j != 0) *(tonari++) = to_pos(i, j - 1);
            if(i != SIZE - 1) *(tonari++) = to_pos(i + 1, j);
            if(j != SIZE - 1) *(tonari++) = to_pos(i, j + 1);
            *tonari = nil;
        }
    }

    LOOP(i, SIZE) {
        LOOP(j, SIZE) {
            tonari = DIAGONALS[to_pos(i, j)];
            if(i != 0 && j != 0)
                *(tonari++) = to_pos(i - 1, j - 1);
            if(i != 0 && j != SIZE - 1)
                *(tonari++) = to_pos(i - 1, j + 1);
            if(i != SIZE - 1 && j != 0)
                *(tonari++) = to_pos(i + 1, j - 1);
            if(i != SIZE - 1 && j != SIZE - 1)
                *(tonari++) = to_pos(i + 1, j + 1);
            *tonari = nil;
        }
    }

    BOARD_ARENA = arena_new(sizeof(Board), 1024);
    return 1;
}

int board_destroy() {
    arena_free(BOARD_ARENA);
    return 1;
}

Board * board_new() {
    Board * self = (Board *)arena_calloc(BOARD_ARENA);
    self->possible_ko = nil;
    return self;
}

void board_free(Board * self) {
    arena_dealloc(BOARD_ARENA, (void *)self);
}

Board * board_clone(Board * self) {
    return (Board *)memcpy(
        arena_malloc(BOARD_ARENA),
        self,
        sizeof(Board)
    );
}

int _search(Pos val, Pos * start, Pos * end) {
    for(; start < end; start ++)
        if (*start == val) return 1;
    return 0;
}

int _has_lib(Color * goban, Pos pos) {
    Pos tonari, *t_ptr;
    Pos arr[SIZE * SIZE];
    Pos * frontier = arr, * end = arr;
    Color color, orig = goban[pos];

    *(end++) = pos;
    frontier = end;

    while(1) {
        ITER(tonari, t_ptr, TONARI[pos]) {
            color = goban[tonari];
            if(color == EMPTY) return 1;
            if(color == orig && !_search(tonari, arr, end))
                *(end++) = tonari;
        }
        if(frontier == end) return 0;
        pos = *(frontier++);
    }
}

int _kill_group(Color * goban, Pos pos) {
    Pos tonari, *t_ptr;
    Pos arr[SIZE * SIZE];
    Pos * frontier = arr, * end = arr;
    Color color, orig = goban[pos];

    int count = 0;
    if (orig == EMPTY) return 0;

    *(end++) = pos;
    frontier = end;

    while(1) {
        goban[pos] = EMPTY;
        count += 1;
        ITER(tonari, t_ptr, TONARI[pos]) {
            color = goban[tonari];
            if(color == orig && !_search(tonari, arr, end))
                *(end++) = tonari;
        }
        if(frontier == end) return count;
        pos = *(frontier++);
    }
}

int board_play(Board * self, Pos pos, Color color) {
    int captures = 0;
    Pos tonari, *t;
    Pos killing[5], *k_ptr = killing;
    Color * goban = self->goban;
    Color opponent = color_other(color);

    if(goban[pos] != EMPTY) return 0;

    goban[pos] = color;

    ITER(tonari, t, TONARI[pos]) {
        if(goban[tonari] == opponent && !_has_lib(goban, tonari))
            *(k_ptr++) = tonari;
    }
    *k_ptr = nil;

    if(k_ptr == killing) { // Empty killing list
        if(_has_lib(goban, pos)) {
            self->possible_ko = nil;
            return 1;
        }
        else {
            goban[pos] = EMPTY;
            return 0;
        }
    }

    ITER(tonari, t, killing)
        captures += _kill_group(goban, tonari);

    if (captures == 1 && self->possible_ko == killing[0]) {
        goban[pos] = EMPTY;
        goban[killing[0]] = opponent;
        return 0;
    }

    self->possible_ko = (captures == 1 ? pos : nil);
    self->captures[color - 1] += captures;

    return 1;
}

int board_print(Board * self, FILE * fh) {
    Color color;
    int i, j;
    LOOP(i, SIZE) {
        LOOP(j, SIZE) {
            color = self->goban[to_pos(i, j)];
            fprintf(fh, "%c ", color == EMPTY ?
                IS_HOSHI(i, j) ? '+' : '.'
            :
                color == BLACK ? 'X' : 'O'
            );
        }
        fprintf(fh, "%d\n", i);
    }
    LOOP(j, SIZE) {
        fprintf(fh, "%d ", j);
    }
    fprintf(fh, "\n");
    fprintf(fh, "X:%d, O:%d+%.1d.5\n",
        self->captures[0],
        self->captures[1],
        KOMI
    );
    return 0;
}
